utils/auction/clear-auction.js

```diff
@@ -0,0 +1,189 @@
+const fs = require('fs');
+const path = require('path');
+
+const USDC_SCALE = 1_000_000n;                // 1 USDC = 1e6
+
+/**
+ * Runs a uniform-price auction using BigInt with:
+ *   - USDC in 6 decimals (1 USDC = 1e6)
+ *   - Tokens in 18 decimals (1 TOKEN = 1e18)
+ *
+ * @param {Object[]} bids
+ * @param {string}   bids[].address      - Ethereum address.
+ * @param {bigint}   bids[].usdcAmount   - USDC in 1e6 units.
+ * @param {bigint}   bids[].maxPrice     - Max price in USDC’s 1e6 format.
+ * @param {bigint}   bids[].priority     - Priority for tie-breaks.
+ * @param {bigint}   totalTokens         - Total tokens to sell in 1e18 units.
+ *
+ * @returns {{
+ *   finalPrice: bigint,  // Price per token in 1e6 USDC units (6 decimals)
+ *   allocations: {
+ *     [address: string]: {
+ *       tokens: bigint,      // Tokens allocated (18 decimals)
+ *       usedUSDC: bigint,    // USDC used (6 decimals)
+ *       refundedUSDC: bigint // USDC refunded (6 decimals)
+ *     }
+ *   }
+ * }}
+ */
+function runAuction(bids, totalTokens) {
+  // Constants for scaling
+  const TOKEN_SCALE = 1_000_000_000_000_000_000n; // 1 TOKEN = 1e18
+
+  // 1) Sort bids by maxPrice (desc), then priority (desc)
+  bids.sort((a, b) => {
+    if (b.maxPrice === a.maxPrice) {
+      return Number(b.priority - a.priority);
+    }
+    return Number(b.maxPrice - a.maxPrice);
+  });
+
+  // 2) Determine the clearing price
+  let tokensAccumulated = 0n;
+  let finalPrice = 0n;
+
+  for (const bid of bids) {
```

Comment:
not sure if this logic is correct.

You are assuming people are getting the tokens at their bid price which is incorrect.

For example I bid 30, you bid 30 and the third person bids 10.

I should be getting my tokens at 10 not at 30, so to calculate tokens accumulated you need to use 10 not 30

Reply:
otherwise we are grabbing not the real last bid, we are finishing up earlier than we should

Reply:
It doesn't assume people are getting the tokens at their bid. The first look is only to calculate how many tokens people are willing to buy.

Reply:
but youa re using this to calculate the finalPrice